### ISSUE TITLE
CA1508: False positive

### ISSUE THREAD

**Issue by HugoRoss** [dotnet/roslyn-analyzers#6947]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/en-gb/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

Not sure how to determine that. I use the following project settings (`*.csproj`) on VS2022 (17.7.3):

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net6.0</TargetFramework>
		<EnableNETAnalyzers>true</EnableNETAnalyzers>
		<AnalysisMode>AllEnabledByDefault</AnalysisMode>
	</PropertyGroup>
</Project>
```

### Describe the bug

In below code, line `if (Object.Equals(default(TOkay), value)) return false;` results in 
`CA1508: 'Object.Equals(default(TOkay), value)' is always 'false'. Remove or refactor the condition(s) to avoid dead code.`

But as you will see when stepping through with the debugger, on the first occasion the condition is not met but on the second it is.

### Steps To Reproduce

Debug through the following code. The cursor walks through both the true- and the false-path of the if-condition.

```
    static void Main(string[] args) {
        bool r1 = IsInvalidCombination(55, new Exception("Whatever")); //Condition is not met, result is true
        bool r2 = IsInvalidCombination(0, new Exception("Whatever")); //Condition is met, result is false
    }

    private static bool IsInvalidCombination<TOkay, TError>(TOkay? value, TError? error) where TError : class {
        if ((value is null) || (error is null)) return false;
        if (Object.Equals(default(TOkay), value)) return false; //<-- Incorrect CA1508!
        return true;
    }
```

### Expected behavior

No CA1508 is shown as the statement is clearly needed and cannot be removed.

### Actual behavior

A CA1508 is shown with the message:
`CA1508: 'Object.Equals(default(TOkay), value)' is always 'false'. Remove or refactor the condition(s) to avoid dead code.`

## Additional context

Windows 10 Enterprise 21H2 19044.3324
Visual Studio 2022 Enterprise 17.7.3
